doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2310 @@
+Given a trait `Trait`, and a type `T` that implements it, we can `as`-cast a reference `&T` into a dynamic trait object: `&dyn Trait`.
+For example:
+```rust
+trait Id {
+    fn get_id(&self) -> usize;
+}
+impl Id for Device { ... }
+
+let device: Device = ...;
+let dyn_id = &device as &dyn Id;  // Create a vtable.
+let id = dyn_id.get_id();  // Indirect procedure call.
+```
+
+`dyn Trait` is a dynamically-sized type, much like slices, and can only exist behind a pointer.
+The reference `&dyn Trait` looks something like this:
+```rust
+struct TraitObject {
+  value: *mut (),
+  vtable: *mut Vtable,
+}
+
+struct Vtable {
+  size: usize,
+  align: usize,
+  dtor: fn(&mut T),
+  // Other trait methods.
+}
+```
+Thus, the dynamic function call to `get_id` would compile to something like the following:
+```rust
+let device: Device = ...;
+let dyn_id = &device as IdTraitObject;
+let id = (dyn_id.vtable.get_id)(dyn_id.value);
+```
+
+There are some limitations on what traits can be made into trait objects: namely, functions cannot take or return functions of type `Self`; only `&Self` or `&mut Self`[^96].
+In other words, all of the functions must treat `Self` as if it were not sized and only accessible through a pointer.[^97] A trait that can be made into a trait object is called _object safe_.
+The type `dyn Trait` always behaves as if it implemented `Trait`, which is relevant for generics, discussed below.
+
+
+#### Unsafe Traits {#unsafe-traits}
+
+It is possible to mark a trait as `unsafe` by writing `unsafe trait MyTrait { .. }`; the only difference with normal traits is that it requires `unsafe impl` to be implemented.
+Unsafe traits typically enforce some kind of additional constraint in addition to their methods; in fact, unsafe traits frequently don't have methods at all.
+For example, the standard library trait `Sync` is implemented by all types which synchronize access[^98].
+Because the invariant this trait asserts is beyond what the compiler can check, it is an unsafe trait.
+
+Trait methods may separately be marked as `unsafe`.
+This is usually done to indicate that not only does care need to be taken in implementing the trait, but calling the function also requires care (and uttering `unsafe`).
+This is separate from marking a trait as `unsafe`, and it is not necessary to mark a trait as `unsafe` for it to have `unsafe` methods.
+
+
+#### Auto Traits
+
+Auto traits are a compiler mechanism for automatically implementing certain traits; in the standard library's source code, this shows up as `auto trait Foo {}` (though this syntax is unavailable for normal libraries).
+Auto traits are implemented automatically for a struct or enum type if all of its fields also implement that trait, and are used for exposing transitive properties to the trait system.
+For example, `Send` and `Sync` are auto traits; a number of other marker traits[^99] are also auto traits.
+
```

Comment:
Perhaps something that explains why some things are Auto Traits versus requiring a manual Derive?

Reply:
Done.